docs/develop/compiler.rst

```diff
@@ -0,0 +1,417 @@
+
+.. code-block:: c
+   :emphasize-lines: 12
+
+   typedef enum _mp_token_kind_t {
+    ...
+    ...
+    P_TOKEN_KW_OR,
+    MP_TOKEN_KW_PASS,
+    MP_TOKEN_KW_RAISE,
+    MP_TOKEN_KW_RETURN,
+    MP_TOKEN_KW_TRY,
+    MP_TOKEN_KW_WHILE,
+    MP_TOKEN_KW_WITH,
+    MP_TOKEN_KW_YIELD,
+    MP_TOKEN_KW_ADD1,
+   } mp_token_kind_t;
+
+Since, we are adding a keyword, edit ``py/lexer.c`` to add the new keyword:
+
+.. code-block:: c
+   :emphasize-lines: 40
+
+   STATIC const char *const tok_kw[] = {
+    "False",
+    "None",
+    "True",
+    "__debug__",
+    "and",
+    "as",
+    "assert",
+    #if MICROPY_PY_ASYNC_AWAIT
+    "async",
+    "await",
+    #endif
+    "break",
+    "class",
+    "continue",
+    "def",
+    "del",
+    "elif",
+    "else",
+    "except",
+    "finally",
+    "for",
+    "from",
+    "global",
+    "if",
+    "import",
+    "in",
+    "is",
+    "lambda",
+    "nonlocal",
+    "not",
+    "or",
+    "pass",
+    "raise",
+    "return",
+    "try",
+    "while",
+    "with",
+    "yield",
+    "add1",
+   };
+
+Notice the keyword is named depending on what you want it to be. For consistency, maintain the
+naming standard accordingly.
+
+.. note::
+   The order of these keywords in ``py/lexer.c`` should match the order of tokens in the enum
+   defined in ``py/lexer.h``.
+
+Parsing
+--------
+
+The parser takes the tokens produced by the lexer and converts them to an abstract syntax tree (AST) or
+*parse tree*. The implementation for the parser is defined in ``py/parse.c``. 
+
+The parser also maintains a table of constants for use in different aspects of parsing, similar to what a `symbol 
+table <https://steemit.com/programming/@drifter1/writing-a-simple-compiler-on-my-own-symbol-table-basic-structure>`_ 
+does.
+
+Several optimizations like `constant folding <http://compileroptimizations.com/category/constant_folding.htm>`_ 
+on integers for all operations i.e logical, binary, unary, etc, optimizing enhancements on parenthesis
+around expressions are performed during this phase and optimizations on strings.
+
+It's worth noting that *docstrings* are discarded and not accessible to the compiler.
+Even optimizations like `string interning <https://en.wikipedia.org/wiki/String_interning>`_ are 
+not applied to *docstrings*.
+
+Compiler passes
+---------------
+
+Like many compilers, MicroPython compiles all code to MicroPython bytecode or native code. The functionality
+that achieves this is implemented in ``py/compile.c``. The most relevant method you should know about is this:
+
+.. code-block:: c
+
+   mp_raw_code_t *mp_compile_to_raw_code(mp_parse_tree_t *parse_tree, 
+   qstr source_file, bool is_repl)
+
+The compiler compiles the code in several passes.
+
+First pass
+~~~~~~~~~~
+
+In the first pass, the compiler computes the stack sizes in scope:
+
+.. code-block:: c
+
+   // compile pass 1
+   comp->emit = emit_bc;
+   #if MICROPY_EMIT_NATIVE
+   comp->emit_method_table = &emit_bc_method_table;
+   #endif
```

Comment:
I think for clarity this #if/#endif pair can be removed (but the code between them left)